src/upgrades/machine_upgrade.py

```diff
@@ -145,21 +145,47 @@ def authorized(self) -> bool:
                 return False
         return False
 
-    def upgrade_unit(self) -> None:
+    def upgrade_unit(self, *, charm) -> None:
         """Runs the upgrade procedure.
 
         Only applies to machine charm.
         """
         logger.debug(f"Upgrading {self.authorized=}")
         self.unit_state = upgrade.UnitState.UPGRADING
 
-        # TODO: Future PR - run mongodb specific checks for upgrade and actually ugprade the snap
-
+        # According to the MongoDB documentation, before upgrading the primary, we must ensure a
+        # safe primary re-election.
+        try:
+            if self._unit.name == charm.primary:
+                logger.debug("Stepping down current primary, before upgrading service...")
+                charm.upgrade.step_down_primary_and_wait_reelection()
+        except mongodb_upgrade.FailedToElectNewPrimaryError:
+            logger.error("Failed to reelect primary before upgrading service.")
+            # todo is it necessary to do something with status which forces the user to run
+            # "force-upgrade"
```

Comment:
@carlcsaposs-canonical - question here 🙏

Reply:
Thank you so much! I really appreciate it! In this case, its actually benefical!